debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java

```diff
@@ -0,0 +1,635 @@
+                optional = true;
+            }
+
+            columns.add(new ColumnMetaData(columnName, postgresType, key, optional, attypmod));
+        }
+
+        PgOutputRelationMetaData metadata = new PgOutputRelationMetaData(relationId, namespace, relationName, columns);
+        if (relationIdMetadata.containsKey(relationId)) {
+            LOGGER.warn("Prior relation message for '{}' has not been consumed by a DML; being replaced.", relationId);
+        }
+
+        TableId tableId = new TableId(null, namespace, relationName);
+
+        LOGGER.trace("Cached relation '{}' with table-id '{}' and synchronizing schema", relationId, tableId);
+
+        relationIdMetadata.put(relationId, metadata);
+        tableIdByRelationId.put(relationId, tableId);
+
+        synchronizeSchemaFromMetadata(metadata);
+    }
+
+    private Map<String, Boolean> resolveColumnOptionality(String schemaName, String tableName) {
+        Map<String, Boolean> columnOptionality = new HashMap<>();
+        try (final PostgresConnection connection = new PostgresConnection(config)) {
+            try {
+                try (ResultSet resultSet = connection.connection().getMetaData().getColumns(null, schemaName, tableName, null)) {
+                    while (resultSet.next()) {
+                        columnOptionality.put(resultSet.getString("COLUMN_NAME"), resultSet.getString("IS_NULLABLE").equals("YES"));
+                    }
+                }
+            }
+            catch (SQLException e) {
+                LOGGER.warn("Failed to read column optionality metadata for '{}.{}'", schemaName, tableName);
+                // todo: DBZ-766 Should this throw the exception or just log the warning?
+            }
+        }
+        return columnOptionality;
+    }
+
+    private void synchronizeSchemaFromMetadata(PgOutputRelationMetaData metadata) throws SQLException {
+        // todo: DBZ-766 a bit of this is based on RecordsStreamProducer#columnValues
+        //      Can we avoid some of the code dupe somehow?
+        try {
+            Table table = schema.tableFor(metadata.getTableId());
+            if (table == null || schemaChanged(metadata, table)) {
+                LOGGER.debug("Schema has changed, applying changes");
+                try (final PostgresConnection connection = new PostgresConnection(config)) {
+                    schema.applySchemaChangesForTable(connection, resolveRelationFromMetadata(metadata));
+                }
+            }
+            else {
+                LOGGER.debug("Schema not changed, no need to apply changes");
+            }
+        }
+        catch (Exception e) {
+            throw new SQLException("Failed to synchronize schema", e);
+        }
+    }
+
+    private boolean schemaChanged(PgOutputRelationMetaData metadata, Table table) {
+        // todo: DBZ-766 a bit of this was borrowed from the RecordsStreamProducer
```

Comment:
+1 for centralizing this. In particular I'm curious why it is needed here, given that the logic from `RecordsStreamProducer` should be run for pgoutput, too?

Reply:
Can you also check out `RecordsStreamProducer#schemaChanged()`, this is still called for each DML message (via `columnValue()`), which it shouldn't, AFAIU. After all, the schema should already be up to date as per the previous "R" event. Perhaps `refreshSchemaIfChanged` should always be `false` in the case of pgoutput?

Reply:
Fixed, this will now be influenced by `ReplicationMessage#shouldSchemaBeSynchronized`.